### ISSUE TITLE
no-empty-pattern - configure behavior for function parameters

### ISSUE THREAD

**Issue by osher** [eslint/eslint#13315]
**What rule do you want to change?**
`no-empty-pattern`

**Does this change cause the rule to produce more or fewer warnings?**
fewer if you use it as is.
more if following the change I'll be able to enable it.

**How will the change be implemented? (New option, new default behavior, etc.)?**
new option

**Please provide some example code that this change will affect:**

Consider a place where an exported/passed method/function must adhere to an arity requirement.
In this case, I use `{}` to communicate that I'm not interested in anything from the argument that is null-destructored - and thus, **_intentionally_** I create no symbols.

A common example - that the error handler of express requires arity of 4.
e.g:
```js
app.use((err, {}, res, {}) => {
   logger.error({ err }, 'got error in route');
   res.headerSent || res.status(500);
   res.end(genericError(err));
});
```
Now I have to disable one of two rules: `no-empty-pattern`, or `no-unused-vars`. Which is a pitty.

**What does the rule currently do for this code?**
fails it.

**What will the rule do after it's changed?**
Provided that I configure it to ignore only the **first-level** destruction of **function arguments** - it will pass, while I can still enjoy errors/warnings about all other cases that might indeed be an error.

**Are you willing to submit a pull request to implement this change?**
I'm totally out of capacity now... I suggest you won't wait for me.

**Comment by anikethsaha** [comment]
Thanks for the Issue.

IMO, this is what the rule is to enforce you to not use `{}`. so I think you can turn off the rule or use something like `app.use((...args) => ...)`. 


let me know if I missed something

**Comment by mdjermanovic** [comment]
@osher thanks for the rule enhancement proposal!

I'm not sure how common this pattern is, since it can't be used for all parameters. In particular, `{}` will throw if it gets `undefined` or `null` argument value.

Is such usage of empty destructuring patterns recommended by a popular style guide?

`no-unused-vars` has [argsIgnorePattern](https://eslint.org/docs/rules/no-unused-vars#argsignorepattern) option to support naming convention for unused parameters (like `_` prefix).

**Comment by osher** [comment]
I'll do my best to explain.
Here's my concrete case.

Consider a DI framework that lets you define your components as factories.
You are **_guarantered_** to be passed with an `options` and a `ioc` bag.

So far this framework is prorietary for an organiztion with a business of about 100 micro-services, all in node-js, but there are talks to open-source the core of the infra, however, I cannot guarantee it now.

Anyway, each one of these micro-services have about 20 modules, most of which are server-side components that are managed by this DI, and therefore, answer to this signature of `module.exports = (options, ioc) => Promise`.

More often than not, the components do require options. Sometimes the options contain settings about how to destruct the `ioc` bag, so they have to come before the `ioc` bag.

In a not too rare case, the components do not need anything from the options.
The way to communicate that is: `module.exports = ({}, ioc) => { ... }`.

If I was not guaranteed an `options`, then I guess the form was `({} = {}, ioc) => { ... }`, but I'm not sure I'd go there.

What I do want to accomplish is to be able to express `({}, { /*destruct ioc*/ })`, and on the same time, be able to rip benefits of bugs like:
`({ optionsBagA: {}}, ioc)` which is definately a bug for `({ optionsBagA = {}}, ioc)`.

Now, the team decided just to turn off this rule, which is a pitty.

**Comment by mdjermanovic** [comment]
Thanks for the additional explanation! Makes sense to me :+1:

**Comment by osher** [comment]
@anikethsaha  - in case like express error-handler - can't use `(...args)` - because it makes the arity of the function to be zero, where by definition - it has to be 4 to be considered an error handler.

Arraity of a function = `func.length`
example:
```
const mw = (req, res, next) => {}
assert.equal(f.length , 3)

const errmw = (err, req, res, next) => {}
assert.equal(f.length , 4)

```
